primary/src/block_synchronizer/responses.rs

```diff
@@ -0,0 +1,135 @@
+use crate::{Certificate, CertificateDigest};
+use blake2::digest::Update;
+use config::Committee;
+use crypto::{traits::VerifyingKey, Digest, Hash};
+use std::fmt::{Display, Formatter};
+use thiserror::Error;
+use tracing::log::{error, warn};
+
+// RequestID helps us identify an incoming request and
+// all the consequent network requests associated with it.
+#[derive(Clone, Debug, Copy, Default, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct RequestID(pub [u8; crypto::DIGEST_LEN]);
+
+impl RequestID {
+    // Create a request key (deterministically) from arbitrary data.
+    pub fn new(data: &[u8]) -> Self {
+        RequestID(crypto::blake2b_256(|hasher| hasher.update(data)))
+    }
+}
+
+impl From<&[CertificateDigest]> for RequestID {
+    fn from(ids: &[CertificateDigest]) -> Self {
+        let mut ids_sorted = ids.to_vec();
+        ids_sorted.sort();
+
+        let result: Vec<u8> = ids_sorted
+            .into_iter()
+            .flat_map(|d| Digest::from(d).to_vec())
+            .collect();
+
+        RequestID::new(&result)
+    }
+}
+
+impl<PublicKey: VerifyingKey> From<&[Certificate<PublicKey>]> for RequestID {
```

Comment:
Consider implementing [`FromIterator`](https://doc.rust-lang.org/std/iter/trait.FromIterator.html) instead, it's more versatile, and you can use it with `collect()`.